pubnub/endpoints/entities/endpoint.py

```diff
@@ -0,0 +1,209 @@
+            if self._uuid_details_level:
+                if self._uuid_details_level == UserIDIncludeEndpoint.USER_ID:
+                    inclusions.append("user_id")
+                elif self._uuid_details_level == UserIDIncludeEndpoint.USER_ID_WITH_CUSTOM:
+                    inclusions.append("user_id.custom")
+
+        if isinstance(self, SpaceIncludeEndpoint):
+            if self._space_details_level:
+                if self._space_details_level == SpaceIncludeEndpoint.CHANNEL:
+                    inclusions.append("space")
+                elif self._space_details_level == SpaceIncludeEndpoint.CHANNEL_WITH_CUSTOM:
+                    inclusions.append("space.custom")
+
+        if isinstance(self, ListEndpoint):
+            if self._filter:
+                params["filter"] = str(self._filter)
+
+            if self._limit:
+                params["limit"] = int(self._limit)
+
+            if self._include_total_count:
+                params["count"] = bool(self._include_total_count)
+
+            if self._sort_keys:
+                joined_sort_params_array = []
+                for sort_key in self._sort_keys:
+                    joined_sort_params_array.append("%s:%s" % (sort_key.key_str(), sort_key.dir_str()))
+
+                params["sort"] = ",".join(joined_sort_params_array)
+
+            if self._page:
+                if isinstance(self._page, Next):
+                    params["start"] = self._page.hash
+                elif isinstance(self._page, Previous):
+                    params["end"] = self._page.hash
+                else:
+                    raise ValueError()
+
+        if len(inclusions) > 0:
+            params["include"] = ",".join(inclusions)
+
+        return params
+
+
+class CustomAwareEndpoint:
+    __metaclass__ = ABCMeta
+
+    def __init__(self):
+        self._custom = None
+
+    def custom(self, custom):
+        self._custom = dict(custom)
+        return self
+
+
+class SpaceEndpoint:
+    __metaclass__ = ABCMeta
+
+    def __init__(self):
+        self._space_id = None
+
+    def space_id(self, space):
+        self._space_id = str(space)
+        return self
+
+    def _validate_space(self):
+        if self._space_id is None or len(self._space_id) == 0:
+            raise PubNubException(pn_error=PNERR_SPACE_MISSING)
+
+
+class UserEndpoint:
+    __metaclass__ = ABCMeta
+
+    def __init__(self):
+        self._user_id = None
+
+    def user_id(self, user_id):
+        self._user_id = str(user_id)
+        return self
+
+    def _effective_user_id(self):
+        if self._user_id is not None:
+            return self._user_id
+        else:
+            return self.pubnub.config.user_id
+
+    def _validate_user_id(self):
+        if self._effective_user_id() is None or len(self._effective_user_id()) == 0:
+            raise PubNubException(pn_error=PNERR_USER_ID_MISSING)
+
+
+class ListEndpoint:
+    __metaclass__ = ABCMeta
+
+    def __init__(self):
+        self._limit = None
```

Comment:
I don't know what is a point of creating "private" objects in class and then create bunch of getters and setters. Even if you add `_` before variable name it still be accessible by everyone, because this is only a gentleman agreement. In such case there is no point to use "private" class variables and then allow users to set then by functions.

Reply:
Well it's all over the SDK and right now i don't want to focus on this